File: example_code/mpi/MPI_Code/Fortran/gatherv.f90
PROGRAM gatherv
USE mpi_f08
IMPLICIT NONE
INTEGER::isend(3)
INTEGER::ircnt(0:2), idisp(0:2)
INTEGER::i,iscnt,myrank
INTEGER,ALLOCATABLE::irecv(:)
ircnt=(/1,2,3/);  idisp=(/0,1,3/)
CALL MPI_Init
CALL MPI_Comm_rank(MPI_COMM_WORLD, myrank)
IF(myrank==0)THEN
  ALLOCATE(irecv(6))
ENDIF
DO i=1,myrank+1
  isend(i)=myrank+1
END DO
iscnt=myrank+1
CALL MPI_Gatherv(isend,iscnt, MPI_INTEGER,irecv,ircnt, idisp,MPI_INTEGER,0,MPI_COMM_WORLD)
IF(myrank==0)THEN
  PRINT*,'irecv=',irecv
ENDIF


irecv=0
DO i=1,myrank+1
  isend(i)=(myrank+1)*10
END DO
IF(myrank==0)THEN
  irecv(1)=isend(1)
  CALL MPI_Gatherv(MPI_IN_PLACE,iscnt,MPI_INTEGER,irecv,ircnt,idisp,MPI_INTEGER,0,MPI_COMM_WORLD)
  PRINT*,'irecv=',irecv
ELSE
  CALL MPI_Gatherv(isend,iscnt,MPI_INTEGER,irecv,ircnt,idisp,MPI_INTEGER,0,MPI_COMM_WORLD)
ENDIF
IF(ALLOCATED(irecv))DEALLOCATE(irecv)
CALL MPI_Finalize
END PROGRAM gatherv

File: example_code/mpi/MPI_Code/C/gatherv.c
#include <stdio.h>
#include <stdlib.h>
#include "mpi.h"
int main()
{
  int i,myrank;
  int isend[3],*irecv;
  int iscnt, ircnt[3]={1,2,3}, idisp[3]={0,1,3};
  MPI_Init(NULL,NULL);
  MPI_Comm_rank(MPI_COMM_WORLD,&myrank);
  if(myrank==0) irecv=(int*)malloc(6*sizeof(int));
  for(i=0;i<myrank+1;i++) isend[i]=myrank+1;
  iscnt=myrank+1;
  MPI_Gatherv(isend, iscnt,MPI_INT, irecv, ircnt,idisp,MPI_INT, 0, MPI_COMM_WORLD);
  if(myrank==0){
    printf("irecv="); for(i=0;i<6;i++) printf("   %d", irecv[i]);
    printf("\n");
  }

  for(i=0;i<myrank+1;i++) isend[i]=(myrank+1)*10;
  if(myrank==0){
    irecv[0]=isend[0];
    MPI_Gatherv(MPI_IN_PLACE,iscnt,MPI_INT,irecv,ircnt,idisp,MPI_INT,0,MPI_COMM_WORLD);
    printf("irecv="); for(i=0;i<6;i++) printf("   %d", irecv[i]);
    printf("\n");
    free(irecv);
  }
  else{
    MPI_Gatherv(isend,iscnt,MPI_INT, irecv,ircnt,idisp,MPI_INT,0,MPI_COMM_WORLD);
  }
  MPI_Finalize();
  return 0;
}
